modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C

```diff
@@ -0,0 +1,215 @@
+
+    // Get the component in the tangential direction
+    _qp_tangential_velocity = relative_velocity * _tangents[_qp][0];
+  }
+  else
+    _qp_tangential_velocity = std::numeric_limits<Real>::quiet_NaN();
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::computeQpIProperties()
+{
+  // Get the _node_to_weighted_gap map
+  ComputeWeightedGapLMMechanicalContact::computeQpIProperties();
+
+  const auto * const node = _lower_secondary_elem->node_ptr(_i);
+  _node_to_weighted_tangential_velocity[node] += _test[_i][_qp] * _qp_tangential_velocity;
+  //  Moose::out << "Node locations: x: " << _lower_secondary_elem->node_ptr(_i)->get_info() <<
+  //  "\n"; Moose::out << "_node_to_weighted_tangential_velocity[node].size(): "
+  //             << _node_to_weighted_tangential_velocity.size() << "\n";
+  //  Moose::out << "qp_tangential_vel: " << MetaPhysicL::raw_value(_qp_tangential_velocity) <<
+  //  "\n"; Moose::out << "_secondary_x_dot[_qp]: " << MetaPhysicL::raw_value(_secondary_x_dot[_qp])
+  //  << "\n";
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::residualSetup()
+{
+  // Clear _node_to_weighted_gap map
+  ComputeWeightedGapLMMechanicalContact::residualSetup();
+
+  _node_to_weighted_tangential_velocity.clear();
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::jacobianSetup()
+{
+  residualSetup();
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::computeResidual(const Moose::MortarType mortar_type)
+{
+  if (mortar_type != Moose::MortarType::Lower)
+    return;
+
+  // ComputeWeightedGapLMMechanicalContact::computeResidual(mortar_type);
+
+  // Fill node to weighted tangential velocity map.
+  for (_qp = 0; _qp < _qrule_msm->n_points(); _qp++)
+  {
+    computeQpProperties();
+    for (_i = 0; _i < _test.size(); ++_i)
+      computeQpIProperties();
+  }
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::computeJacobian(const Moose::MortarType mortar_type)
+{
+  // During "computeResidual" and "computeJacobian" we are actually just computing properties on the
+  // mortar segment element mesh. We are *not* actually assembling into the residual/Jacobian. For
+  // the zero-penetration constraint, the property of interest is the map from node to weighted gap.
+  // Computation of the properties proceeds identically for residual and Jacobian evaluation hence
+  // why we simply call computeResidual here. We will assemble into the residual/Jacobian later from
+  // the post() method
+  computeResidual(mortar_type);
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::post()
+{
+  // Enforce weighted gap constraint (calls enforceConstraintOnNode)
+  ComputeWeightedGapLMMechanicalContact::post();
+
+  // Enforce weighted tangential velocity constraint
+  for (const auto & pr : _node_to_weighted_tangential_velocity)
+  {
+    _weighted_gap_ptr = &_node_to_weighted_gap[pr.first];
+    _tangential_vel_ptr = &pr.second;
+    enforceConstraintOnNode(pr.first);
+  }
```

Comment:
So here you are going to end up doing two loops. One from `ComputeWeightedGapLMMechanicalContact::post` and then one from here. I don't think you should call `ComputeWeightedGapLMMechanicalContact::post` here. Instead you should uncomment the `Compute::WeightedGapLMMechanicalContact::enforceConstraintOnNode` call in your `enforceConstraintOnNode` override

Reply:
Makes sense